### DISCUSSION TITLE
Connection refused for NextJS inside Docker swarm

### DISCUSSION THREAD

**Discussion by indapublic** [vercel/next.js#55901] [Help] [answered]
### Summary

I have my private self-hosted docker image with nextjs inside. This image exposes 3000 port.

I tried to run this image via docker and docker swarm.

1. **Docker**

```
docker run -p 3000:3000 path-to-my-image
```

**Result:**

`docker ps`:

```
0.0.0.0: 3000->3000/tcp, ::: 3000->3000/tcp
```

`ps` inside container:

```
PID   USER     TIME  COMMAND
    1 nextjs    0:00 next-router-wo
   22 nextjs    0:01 next-render-worker-app
   23 nextjs    0:01 next-render-worker-pages
   51 nextjs    0:00 sh
   57 nextjs    0:00 ps
```

`netstat -tulp|grep 3000` inside container:

```
tcp        0      0 cf748210f6f3:3000       0.0.0.0:*               LISTEN      1/next-router-wo
```

`curl http://127.0.0.1:3000` outputs correct response. So all is fine.

2. **Docker swarm**

`stack.yaml`

```
version: "3.7"

services:
  app:
    image: path-to-my-image
    ports:
      - 3000:3000
    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints: [node.role == manager]
```

`docker service ls` shows service is online, 1 replica

`ps` inside container:

```
PID   USER     TIME  COMMAND
    1 nextjs    0:00 next-router-wo
   22 nextjs    0:01 next-render-worker-app
   23 nextjs    0:01 next-render-worker-pages
   50 nextjs    0:00 sh
   56 nextjs    0:00 ps
```

`netstat -tulp|grep 3000` inside container:

```
tcp        0      0 3e53ca570e77:3000       0.0.0.0:*               LISTEN      1/next-router-wo
```

But `curl http://127.0.0.1:3000` fails:

```
curl: (7) Failed to connect to 127.0.0.1 port 3000 after 0 ms: Connection refused
```

Tried another port, another machines, tried `node` and `pm2`. All is same.

I can’t understand the reason and where to dig next.

Thank you in advance.

### Additional information

```js
Did two repositories (and public docker images): 

1. github.com/indapublic/docker-swarm-expose-test (plain js, works) 

2. github.com/indapublic/docker-swarm-expose-test-nextjs (next js, doesn't work)
```

**Comment by NolanTrem** [comment, marked as answer]
By explicitly setting `HOSTNAME` in my next.js project's Docker image, I was able to solve this.

I found that when I run:
```
docker run -p 3000:3000 myimage
```
- Docker maps the host machine’s port 3000 directly to the container’s port 3000
- If Next.js is listening on `0.0.0.0` inside the container, then from your host machine, you can hit `http://127.0.0.1:3000` (or `http://localhost:3000`) successfully

With Docker Swarm when you do:
```yaml
version: "3.7"
services:
  app:
    image: myimage
    ports:
      - "3000:3000"
    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints: [node.role == manager]
```

Then run:

```
docker stack deploy -c stack.yaml myapp
```

- I see the container is running and is listening on port 3000 (inside the container)
- But accessing `http://127.0.0.1:3000` (or `curl localhost:3000`) from the manager node gives “connection refused” or "connection reset"

This seems to come to the fact that Swarm publishes that port via its ingress network , so you typically have to use the node’s actual IP (e.g., `192.168.x.x:3000`) from outside the container or from another machine. 

Also, if your Next.js server is **not** listening on `0.0.0.0`, but only on `localhost`, that can cause issues in Swarm. The container needs to bind to all interfaces so Swarm’s ingress traffic can route through.

I think this comes down to two main things:
1. Swarm Routing Mesh
   - When you expose a port in Swarm (`3000:3000`), Swarm sets up routing that listens on **the host IP** (generally `0.0.0.0` on the node).  
   - Traffic arrives at that external IP:port combination, then is forwarded inside to the container’s network interface on port 3000.  
   - If you try to `curl 127.0.0.1:3000` inside the container, it’s not the same interface that’s receiving the published traffic from outside.  

2. Next.js May Default to `localhost`
   - Some Node or Next.js setups default to listening on `127.0.0.1` or `localhost` only.  
   - For container-based deployment (especially in Swarm), you typically need to bind to `0.0.0.0`. That ensures the service is reachable on any network interface, including the swarm overlay network.  
   - If Next.js only listens on `127.0.0.1`, connections from the ingress are dropped.

TLDR; Set `HOSTNAME=0.0.0.0` (or use `-H 0.0.0.0` in your startup command) and then curl your node’s IP, you should see a successful response.

**Reply by jingjiezhou2021** [reply]
P.S. Even if you pass `-H 0.0.0.0`, Docker automatically sets a `HOSTNAME` environment variable (the container ID). Next.js will sometimes pick up this `HOSTNAME` and bind to that internal container IP instead of `0.0.0.0`.
The practical fix is always explicitly set the environment variable `HOSTNAME=0.0.0.0` in your `Dockerfile` or compose file

**Comment by riteshreg** [comment]
facing same issue
